moveit_experimental/jog_arm/src/jog_arm/jog_calcs.cpp

```diff
@@ -0,0 +1,763 @@
+///////////////////////////////////////////////////////////////////////////////
+//      Title     : jog_calcs.h
+//      Project   : jog_arm
+//      Created   : 1/11/2019
+//      Author    : Brian O'Neil, Andy Zelenak, Blake Anderson
+//
+// BSD 3-Clause License
+//
+// Copyright (c) 2019, Los Alamos National Security, LLC
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <jog_arm/jog_calcs.h>
+
+namespace jog_arm
+{
+// Constructor for the class that handles jogging calculations
+JogCalcs::JogCalcs(const JogArmParameters parameters, JogArmShared& shared_variables, pthread_mutex_t& mutex,
+                   const robot_model_loader::RobotModelLoaderPtr model_loader_ptr)
+  : move_group_(parameters.move_group_name), tf_listener_(tf_buffer_), parameters_(parameters)
+{
+  // Publish collision status
+  warning_pub_ = nh_.advertise<std_msgs::Bool>(parameters_.warning_topic, 1);
+
+  // MoveIt! Setup
+  while (ros::ok() && !model_loader_ptr)
+  {
+    ROS_WARN_THROTTLE_NAMED(5, LOGNAME, "Waiting for a non-null robot_model_loader pointer");
+    ros::Duration(WHILE_LOOP_WAIT).sleep();
```

Comment:
> Why don't you use the yaml-parameter publish_period here?

I think publish_period should only be related to the rate of message publication. I would rather keep a separate "wait" parameter here, since it's not related to publishing.

> If I set publish_period much higher than WHILE_LOOP_WAIT, I loose the stop message in velocity mode and the manipulator moves on...

This seems like a different issue that is definitely worth me digging into.

Reply:
cba6927 should take care of your "runaway robot" issue.

Reply:
From a Sense->Reason->Act perspective I don't understand your threading architecture here. Can you elaborate more on that (maybe add some documentation explaining the architecture inside the code?)

I'd expect the calculation to be necessary once per `publish_period` more often is a waste of cpu power, slower would publish old values. Therefore I'd suggest merging the threads or at run them with the same rate.

Reply:
Originally, the jogging calculations were very slow (~40 Hz) so it made sense to run that thread as fast as possible. After some optimization it is running much faster now, ~500-1000 Hz, so I think it does make sense to match rates, like you said.

Reply:
And CPU load drops from ~21-24% to 15-17%. Nice!